cat-home/src/test/java/com/dianping/cat/report/task/TaskHelperTest.java

```diff
@@ -24,45 +24,170 @@
+        cal.set(Calendar.DAY_OF_YEAR, 1);
+        cal.set(Calendar.HOUR_OF_DAY, 0);
+        cal.set(Calendar.MINUTE, 0);
+        cal.set(Calendar.SECOND, 0);
+        cal.set(Calendar.MILLISECOND, 0);
+        Date yesterdayZero = TaskHelper.yesterdayZero(cal.getTime());
+        Assert.assertEquals(cal.getTimeInMillis(),
+                yesterdayZero.getTime() + 3600L * 1000 * 24);
+    }
+
+    @Test
+    public void testYesterdayZero2() {
+        Assert.assertEquals(new Date(1562626800000L),
+                TaskHelper.yesterdayZero(new Date(1562769422000L)));
+    }
+
+    @Test
+    public void testTodayZero() {
+        Calendar cal = Calendar.getInstance();
+        cal.set(Calendar.DAY_OF_YEAR, 1);
+        cal.set(Calendar.HOUR_OF_DAY, 0);
+        cal.set(Calendar.MINUTE, 0);
+        cal.set(Calendar.SECOND, 0);
+        cal.set(Calendar.MILLISECOND, 1);
+        Date todayZero = TaskHelper.todayZero(cal.getTime());
+        Assert.assertEquals(cal.getTimeInMillis(), todayZero.getTime() + 1L);
+    }
+
+    @PrepareForTest({TaskHelper.class})
+    @Test
+    public void testTodayZero2() throws Exception {
+        Date date = new Date(1562799600000L);
+        PowerMockito.whenNew(Date.class).withAnyArguments().thenReturn(date);
+        Assert.assertEquals(date,
+                TaskHelper.todayZero(null));
+    }
+
+    @Test
+    public void testTodayZero3() {
+        Assert.assertEquals(new Date(1562626800000L),
+                TaskHelper.todayZero(new Date(1562683022000L)));
+    }
+
+    @Test
+    public void testJoinIntArray() {
+        final int[] array = {1, -8, -7, -7, -7, -7, -7, -253_968, -7, -24, -24,
+                -2147475464, 4088, -2147483647, 9, 33505273, 1, -8};
+
+        Assert.assertNull(TaskHelper.join((int[]) null, '\u0000',
+                -160729767, -159952269));
+
+        Assert.assertEquals("",
+                TaskHelper.join(new int[0], '\u0000'));
+        Assert.assertEquals("1\u0000-8",
+                TaskHelper.join(array, '\u0000', 16, 17));
+        Assert.assertEquals("",
+                TaskHelper.join(new int[0], '\u0000',
+                        -2147483648, 0));
+        Assert.assertEquals("",
+                TaskHelper.join(new int[0], '\u0000',
+                        1816002434, -1402208264));
+    }
+
+    public void testJoinDoubleArray() {
+        Assert.assertNull(TaskHelper.join((double[]) null, '\u0000',
+                -402653175, 1669333001));
+
+        Assert.assertEquals("",
+                TaskHelper.join(new double[]{5.1}, '\u0000'));
+        Assert.assertEquals("2.1\u00004.5",
+                TaskHelper.join(new double[]{2.1, 4.5}, '\u0000'));
+        Assert.assertEquals("",
+                TaskHelper.join(new double[]{5.1, 3.5}, '\u0000',
+                        -450887681, -1299185664));
+        Assert.assertEquals("",
+                TaskHelper.join(new double[]{5.1, 9.2}, '\u0000',
+                        1197600769, 964623358));
```

Comment:
I would use smaller and readable values as `startIndex` and `endIndex` check is `if (bufSize <= 0)`

Reply:
done